### ISSUE TITLE
2.10: URL types issues (application of constraints)

### ISSUE THREAD

**Issue by dhofstetter** [pydantic/pydantic#10914]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Pydantic v2.10 has issues with Url types and AfterValidators (see example below)

```
tests/security/keycloak/test_config.py:None (tests/security/keycloak/test_config.py)
tests/security/keycloak/test_config.py:5: in <module>
    from yio_fastapi_utils.security.keycloak import KeycloakSettings
yio_fastapi_utils/security/keycloak/__init__.py:5: in <module>
    from yio_fastapi_utils.security.keycloak.deps import security_dependency, AccessToken
yio_fastapi_utils/security/keycloak/deps.py:12: in <module>
    from yio_fastapi_utils.security.keycloak.config import KeycloakSettings
yio_fastapi_utils/security/keycloak/config.py:17: in <module>
    class KeycloakSettings(BaseSettings):
venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:226: in __new__
    complete_model_class(
venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py:658: in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
venv/lib/python3.11/site-packages/pydantic/main.py:697: in __get_pydantic_core_schema__
    return handler(source)
venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:84: in __call__
    schema = self._handler(source_type)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:612: in generate_schema
    schema = self._generate_schema_inner(obj)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:881: in _generate_schema_inner
    return self._model_schema(obj)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:693: in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:693: in <dictcomp>
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1073: in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:1261: in _common_field_schema
    schema = self._apply_annotations(
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2051: in _apply_annotations
    schema = get_inner_schema(source_type)
venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:84: in __call__
    schema = self._handler(source_type)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2126: in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
venv/lib/python3.11/site-packages/pydantic/functional_validators.py:73: in __get_pydantic_core_schema__
    schema = handler(source_type)
venv/lib/python3.11/site-packages/pydantic/_internal/_schema_generation_shared.py:84: in __call__
    schema = self._handler(source_type)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2127: in new_handler
    schema = self._apply_single_annotation(schema, annotation)
venv/lib/python3.11/site-packages/pydantic/_internal/_generate_schema.py:2093: in _apply_single_annotation
    maybe_updated_schema = _known_annotated_metadata.apply_known_metadata(metadata, schema.copy())
venv/lib/python3.11/site-packages/pydantic/_internal/_known_annotated_metadata.py:285: in apply_known_metadata
    raise RuntimeError(f"Unable to apply constraint '{constraint}' to schema of type '{schema_type}'")
E   RuntimeError: Unable to apply constraint 'allowed_schemes' to schema of type 'function-wrap'
```

### Example Code

```Python
def remove_trailing_slash(url: str) -> str:
    """
    Remove trailing slash from url.

    :param url: url to be cleaned
    :return: cleaned url
    """
    return url[:-1] if url.endswith("/") else url


FQHttpsUrl = Annotated[AnyUrl, UrlConstraints(allowed_schemes=['https'])]

RealmName = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, to_lower=True)]

ServerUrl = Annotated[FQHttpsUrl, AfterValidator(lambda value: FQHttpsUrl(remove_trailing_slash(str(value))))]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.0
        pydantic-core version: 2.27.0
          pydantic-core build: profile=release pgo=false
                 install path: /Users/dh/Development/yio/yio/yio-fastapi-utils/venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.10 (main, Sep  7 2024, 01:03:31) [Clang 15.0.0 (clang-1500.3.9.4)]
                     platform: macOS-15.0-arm64-arm-64bit
             related packages: pydantic-settings-2.6.1 pydantic-extra-types-2.10.0 mypy-1.8.0 fastapi-0.115.5 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by dhofstetter** [comment]
I'm pretty sure now that the issue comes down to this construct:

```python
from typing import Annotated

from pydantic import AnyUrl, UrlConstraints, StringConstraints, AfterValidator, TypeAdapter
from pydantic.experimental.pipeline import validate_as

# not working anymore
FQHttpsUrlBroken = Annotated[AnyUrl, UrlConstraints(allowed_schemes=['https'])]

TypeAdapter(FQHttpsUrlBroken)
```

**Comment by sydney-runkle** [comment]
@dhofstetter,

Thanks for reporting this! We're taking a closer look now!

**Comment by sydney-runkle** [comment]
I'll note - we've touched on some [changes to urls](https://pydantic.dev/articles/pydantic-v2-10-release#migrate-to-subclassing-instead-of-annotated-approach-for-pydantic-url-types) in v2.10 in our blog post.

The above should still work. That being said, we've migrated away form this `Annotated` approach, given that initializing annotated types isn't supported, and bare annotated types fail to enforce constraints (unlike the new subclass oriented url types).

**Comment by sydney-runkle** [comment]
I've opened https://github.com/pydantic/pydantic/pull/10922 to fix this issue :)

**Comment by antazoey** [comment]
I am also having maybe similar issue, I cannot use FileUrl for AnyUrl anymore on 2.10

**Comment by sydney-runkle** [comment]
@antazoey, I've replied on #10922 :)

**Comment by sydney-runkle** [comment]
Hey folks! We've just released [v2.10.1](https://github.com/pydantic/pydantic/releases/tag/v2.10.1) with a fix for this issue!

Let us know if you're still experiencing any difficulties. Thanks!
